### DISCUSSION TITLE
How to force Babel to preserve escape characters during transformation?

### DISCUSSION THREAD

**Discussion by gogo9th** [babel/babel#13104] [General] [answered]
Hi,

I am using Babel to transform Javascript used in webpages, but I have an issue in transforming the following code: 

`var a = "\u003C\u002Fscript\u003E";`

. . . because Babel transforms the above as: 

`var a = "</script>";`

This is problematic, because a web browser's JavaScript engine recognizes the substring `"</script>` as the end of the script tag. How can I set the Babel option to keep the original string as it is (`"\u003C\u002Fscript\u003E"`) even after transformation?

I would really appreciate the exact syntax to enforce this, because I mostly find it hard to find the correct syntax... Thanks again!

```
      var options = {
        retainLines: true,
        compact: false,
        sourceType: 'unambiguous',
        sourceMaps: true,
        presets: [ ['@babel/preset-env',
                     { "exclude": ["@babel/plugin-transform-function-name"] }
                  ] ],
        plugins: [
          ["transform-es2015-modules-commonjs-simple",
            {
              'noMangle': true,
              'allowTopLevelThis': true,
              'strictMode': false
            },
          ]
        ]
     }
```

**Comment by gogo9th** [comment]
I tried the following (trying to exclude `@babel/plugin-transform-unicode-escapes` and `@babel/plugin-transform-unicode-regex`), but still not working:

```
      var options = {
        retainLines: true,
        compact: false,
        sourceType: 'unambiguous',
        sourceMaps: true,
        presets: [ ['@babel/preset-env', 
                     { "exclude": [ "@babel/plugin-transform-function-name",
                                    '@babel/plugin-transform-unicode-escapes',
                                    '@babel/plugin-transform-unicode-regex'
                                 ] }
                  ] ],
        plugins: [
          ["transform-es2015-modules-commonjs-simple",
            {     
              'noMangle': true,
              'allowTopLevelThis': true,
              'strictMode': false
            },
          ]
        ]
      }
```

**Comment by JLHwung** [comment, marked as answer]
You can use `generateOpts` to control how Babel should generate your code based on the transformed AST:
```js
var options = {
  generatorOpts: {
    jsescOtpion: {
      isScriptContext: true
    }
  }
}
```

https://babeljs.io/docs/en/babel-generator#options
https://github.com/mathiasbynens/jsesc#isscriptcontext

**Reply by gogo9th** [reply]
Sorry, I tried this option, but still does not work:

```
      var options = {               
        generatorOpts: {
          jsescOtpion: {
            isScriptContext: true
          }
        },
        retainLines: true,
        compact: false,
        sourceType: 'unambiguous',
        sourceMaps: true,
        presets: [ ['@babel/preset-env',
                     { "exclude": [ "@babel/plugin-transform-function-name",
                                    '@babel/plugin-transform-unicode-escapes',
                                    '@babel/plugin-transform-unicode-regex'
                                 ] }
                  ] ],
      }
```

**Reply by gogo9th** [reply]
In general, I am looking for a way to disable all kinds of unicode transformation (including handling </script>).

**Reply by nicolo-ribaudo** [reply]
Unicode transformations in strings are handled by `@babel/plugin-transform-literals`, but it also handles `0o` and `0b` number literals.

**Reply by gogo9th** [reply]
Thanks, excluding this works!
```      
      var options = {
        retainLines: true,
        compact: false,
        sourceType: 'unambiguous',
        sourceMaps: true,
        presets: [ ['@babel/preset-env',
                     { "exclude": [ "@babel/plugin-transform-function-name",
                                          '@babel/plugin-transform-literals',
                                 ] }
                  ] ],
        plugins: [
          ["transform-es2015-modules-commonjs-simple",
            {
              'noMangle': true,
              'allowTopLevelThis': true,
              'strictMode': false
            },
          ]
        ]
      }
```

**Reply by nicolo-ribaudo** [reply]
Yeah but keep in mind that by doing it some code might not work in some very old browsers that need that transform.

**Reply by gogo9th** [reply]
Basically, I'm transforming the original Javascript code that is to be run by a browser. Therefore, if some unicode is incompatible with a browser, then the original Javascript code it would have run will also be incompatible- this means our transformation does not introduce an additional problem. So I believe it should be fine.
